lib/broadway_cloud_pub_sub/pull_client.ex

```diff
@@ -0,0 +1,170 @@
+
+    config
+    |> execute(:pull, %{"maxMessages" => max_messages})
+    |> Map.fetch!("receivedMessages")
+    |> wrap_received_messages(ack_builder)
+  end
+
+  @impl Client
+  def put_deadline(ack_ids, ack_deadline_seconds, config) do
+    payload = %{
+      "ackIds" => ack_ids,
+      "ackDeadlineSeconds" => ack_deadline_seconds
+    }
+
+    execute(config, "modifyAckDeadline", payload)
+
+    :ok
+  end
+
+  @impl Client
+  def acknowledge(ack_ids, config) do
+    execute(config, :acknowledge, %{"ackIds" => ack_ids})
+  end
+
+  defp wrap_received_messages(pub_sub_messages, ack_builder) do
+    Enum.map(pub_sub_messages, fn pub_sub_msg ->
+      pub_sub_msg_to_broadway_msg(pub_sub_msg, ack_builder)
+    end)
+  end
+
+  defp pub_sub_msg_to_broadway_msg(pub_sub_msg, ack_builder) do
+    %{
+      "ackId" => ack_id,
+      "message" => message
+    } = pub_sub_msg
+
+    {data, metadata} =
+      message
+      |> decode_message()
+      |> Map.pop("data")
+
+    metadata = %{
+      attributes: metadata["attributes"],
+      messageId: metadata["messageId"],
+      publishTime: parse_datetime(metadata["publishTime"])
+    }
+
+    %Message{
+      data: data,
+      metadata: metadata,
+      acknowledger: ack_builder.(ack_id)
+    }
+  end
+
+  defp parse_datetime(nil), do: nil
+
+  defp parse_datetime(str) when is_binary(str) do
+    case DateTime.from_iso8601(str) do
+      {:ok, dt, _} ->
+        dt
+
+      err ->
+        raise """
+        invalid datetime string: #{inspect(err)}
+        """
+    end
+  end
+
+  defp decode_message(%{"data" => nil} = message), do: message
+
+  defp decode_message(%{"data" => encoded_data} = message) do
+    %{message | "data" => Base.decode64!(encoded_data)}
+  end
+
+  defp headers(config) do
+    token = get_token(config)
+    [{"authorization", "Bearer #{token}"}, {"content-type", "application/json"}]
+  end
+
+  defp url(config, action) do
+    sub = URI.encode(config.subscription.string)
+    path = "/v1/" <> sub <> ":" <> to_string(action)
+    config.base_url <> path
+  end
+
+  defp status!(%Response{status: same} = resp, same) do
+    resp
+  end
+
+  defp status!(%Response{status: got}, expected) do
+    raise "#{inspect(__MODULE__)} - unexpected HTTP status code -" <>
+            " expected: #{expected}, got: #{inspect(got)}"
+  end
+
+  defp json!(%Response{body: b}), do: Jason.decode!(b)
+
+  defp execute(config, action, payload) do
+    url = url(config, action)
+    body = Jason.encode!(payload)
+    token = get_token(config)
+    headers = headers(token)
+
+    config.finch_name
+    |> finch_request!(url, body, headers)
+    |> status!(200)
+    |> json!()
+  end
+
+  defp finch_request!(finch_name, url, body, headers) do
+    :post
+    |> Finch.build(url, headers, body)
+    |> Finch.request(finch_name, receive_timeout: :infinity)
+    |> case do
+      {:ok, resp} ->
+        resp
+
+      {:error, %{__exception__: true} = err} ->
+        raise err
```

Comment:
```suggestion
      {:error, err} ->
        raise err
```

Reply:
We ended up changing this to match what the GoogleApiClient did. HTTP error cases and bad statuses are logged instead of raising.